examples/common.cpp

```diff
@@ -60,7 +53,71 @@ int32_t get_num_physical_cores() {
         return num_physical_cores;
     }
 #elif defined(_WIN32)
-    //TODO: Implement
+    SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX *buffer = nullptr;
+    DWORD length = 0;
+
+    // Call GetLogicalProcessorInformationEx with a nullptr buffer to get the required buffer length
+    GetLogicalProcessorInformationEx(RelationAll, nullptr, &length);
+
+    // Allocate memory for the buffer
+    buffer = static_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX *>(malloc(length));
+
+    // Things to count
+    unsigned int physical_cores = 0;
+    unsigned int physical_performance_cores = 0;
+    unsigned int physical_efficiency_cores = 0;
+    unsigned int logical_cores = 0;
+    unsigned int logical_performance_cores = 0;
+    unsigned int logical_efficiency_cores = 0;
+
+    // Call GetLogicalProcessorInformationEx again with the allocated buffer
+    if (GetLogicalProcessorInformationEx(RelationAll, buffer, &length)) {
+        DWORD offset = 0;
+
+        while (offset < length) {
+            auto info = reinterpret_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX *>(reinterpret_cast<char *>(buffer) + offset);
+
+            if (info->Relationship == RelationProcessorCore) {
+                physical_cores += info->Processor.GroupCount;
+
+                for (WORD i = 0; i < info->Processor.GroupCount; ++i) {
+                    int core_count = static_cast<int>(__popcnt64(info->Processor.GroupMask[i].Mask));
```

Comment:
Will `__popcnt64` work on other compilers such as mingw? Sadly, `std::popcount` is C++20.

Reply:
I don't know about mingw. It sounds like it didn't work on older versions of it anyway.

`__popcnt64` is used here for counting the logical cores, which I believe to be unnecessary. I can rip out all the logical core stuff if it's working on P/E systems since we only care about the physical cores count on modern CPUs. I have yet to see a single case where hyperthreading helps llama.cpp performance.